internal/server/singleprocess/state/snapshot.go

```diff
@@ -0,0 +1,462 @@
+			},
+		}); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// StageRestoreSnapshot stages a database restore for the next server restart.
+// This will create a temporary file alongside the data file so we must have
+// write access to the directory containing the database.
+func (s *State) StageRestoreSnapshot(r io.Reader) error {
+	log := s.log.Named("restore")
+	log.Warn("beginning to stage snapshot restore")
+
+	ri := newRestoreInfo(log, s.db)
+	if err := ri.Lock(); err != nil {
+		return err
+	}
+	defer ri.Unlock()
+
+	// Get our file info
+	fi, err := os.Stat(s.db.Path())
+	if err != nil {
+		return err
+	}
+
+	// Open our temporary file and copy the restore contents into it.
+	log.Info("copying the snapshot data to a temporary path", "path", ri.StageTempPath)
+	tempF, err := os.OpenFile(ri.StageTempPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fi.Mode())
+	if err != nil {
+		log.Error("error creating temporary path", "err", err)
+		return err
+	}
+	n, err := io.Copy(tempF, r)
+	log.Info("copied snapshot data", "len", n)
+	tempF.Close()
+	if err != nil {
+		log.Error("error copying snapshot data", "err", err)
+		return err
+	}
+
+	// We do want to check that this appears valid because if we can't open
+	// the file then the server will never start again until it is cleaned up.
+	log.Info("validating restore file")
+	sr, closer, err := snapshotReader(ri.StageTempPath, nil)
+	if err != nil {
+		log.Error("error while validating restore file", "err", err)
+		return fmt.Errorf("error validating restore data: %s", err)
+	}
+
+	var header pb.Snapshot_Header
+	err = sr.ReadMsg(&header)
+	if cerr := closer(); cerr != nil {
+		err = cerr
+	}
+	if err != nil {
+		log.Error("error while validating restore file", "err", err)
+		return fmt.Errorf("error validating restore data: %s", err)
+	}
+
+	// Replace our file
+	log.Info("atomically replacing file", "src", ri.StageTempPath, "dest", ri.StagePath)
+	if err := atomic.ReplaceFile(ri.StageTempPath, ri.StagePath); err != nil {
+		log.Error("error replacing file", "err", err)
+		return err
+	}
+
+	// Open the new database
+	log.Warn("snapshot staged for restore on next startup")
+	return nil
+}
+
+// finalizeRestore checks for any staged restore and moves it into place.
+// This will error if it fails for any reason which may prevent startup
+// but we have to prevent startup because the user wanted a restore.
+func finalizeRestore(log hclog.Logger, db *bolt.DB) (*bolt.DB, error) {
+	log.Debug("checking if DB restore is requested")
+	ri := newRestoreInfo(log, db)
+	if err := ri.Lock(); err != nil {
+		return db, err
+	}
+	defer ri.Unlock()
+
+	_, err := os.Stat(ri.StagePath)
+	if os.IsNotExist(err) {
+		log.Debug("no restore file found, no DB restore requested")
+		return db, nil
+	}
+	if err != nil {
+		log.Error("error checking for restore file", "err", err)
+		return db, err
+	}
+
+	log.Warn("restore file found, will initiate database restore")
```

Comment:
Is it worth being extra explicit here to include the file path of the restore file?

```suggestion
	log.Warn("restore file found, will initiate database restore", "path", ri.StagePath)
```